File: android/src/main/java/expo/modules/superwallexpo/SuperwallExpoModule.kt
package expo.modules.superwallexpo

import android.content.pm.PackageManager
import android.net.Uri
import android.app.Application
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.Promise
import expo.modules.superwallexpo.bridges.PurchaseControllerBridge
import expo.modules.superwallexpo.bridges.SuperwallDelegateBridge
import expo.modules.superwallexpo.json.*
import com.superwall.sdk.Superwall
import com.superwall.sdk.delegate.SuperwallDelegate
import com.superwall.sdk.delegate.PurchaseResult
import com.superwall.sdk.delegate.RestorationResult
import com.superwall.sdk.models.entitlements.SubscriptionStatus
import com.superwall.sdk.models.entitlements.Entitlement
import com.superwall.sdk.identity.IdentityOptions
import com.superwall.sdk.identity.identify
import com.superwall.sdk.identity.setUserAttributes
import com.superwall.sdk.config.options.SuperwallOptions
import com.superwall.sdk.config.options.EventTrackingBehavior
import com.superwall.sdk.logger.LogLevel
import com.superwall.sdk.network.device.InterfaceStyle
import com.superwall.sdk.paywall.presentation.PaywallPresentationHandler
import com.superwall.sdk.paywall.presentation.CustomCallbackResult
import com.superwall.sdk.paywall.presentation.result.PresentationResult
import com.superwall.sdk.config.models.ConfigurationStatus
import com.superwall.sdk.paywall.presentation.register
import com.superwall.sdk.paywall.presentation.dismiss
import com.superwall.sdk.paywall.view.webview.PaywallResource
import com.superwall.sdk.paywall.presentation.get_presentation_result.getPresentationResult
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.future.await
import kotlinx.coroutines.runBlocking
import android.util.Log
import java.util.Collections
import java.util.WeakHashMap
import java.util.concurrent.CompletableFuture

class SuperwallExpoModule : Module() {

  companion object {
    var instance: SuperwallExpoModule? = null
    private val instances = Collections.newSetFromMap(WeakHashMap<SuperwallExpoModule, Boolean>())

    // Events must reach every live module instance. More than one app context can
    // exist at once (e.g. expo-dev-client's launcher plus the app), and `instance`
    // may point at a module whose JS runtime never subscribed - sending only
    // through it silently drops events like `onPurchase`, leaving the native
    // purchase future incomplete forever.
    fun emitEvent(name: String, body: Map<String, Any?>?) {
      val liveInstances = synchronized(instances) { instances.toList() }
      liveInstances.forEach { it.sendEvent(name, body ?: emptyMap()) }
    }
  }

  val scope = CoroutineScope(Dispatchers.Main)
  val ioScope = CoroutineScope(Dispatchers.IO)
  var backPressedPromise: CompletableFuture<Boolean>? = null

  init {
    instance = this
    synchronized(instances) { instances.add(this) }
  }

  private val onPaywallPresent = "onPaywallPresent"
  private val onPaywallDismiss = "onPaywallDismiss"
  private val onPaywallError = "onPaywallError"
  private val onPaywallSkip = "onPaywallSkip"
  private val onCustomCallback = "onCustomCallback"

  val customCallbackFutures = java.util.concurrent.ConcurrentHashMap<String, CompletableFuture<CustomCallbackResult>>()

  // Purchase Events
  private val onPurchase = "onPurchase"
  private val onPurchaseRestore = "onPurchaseRestore"

  // Back Button Event
  private val onBackPressed = "onBackPressed"

  // Legacy Events
  private val subscriptionStatusDidChange = "subscriptionStatusDidChange"
  private val handleSuperwallEvent = "handleSuperwallEvent"
  private val handleCustomPaywallAction = "handleCustomPaywallAction"
  private val willDismissPaywall = "willDismissPaywall"
  private val willPresentPaywall = "willPresentPaywall"
  private val didDismissPaywall = "didDismissPaywall"
  private val didPresentPaywall = "didPresentPaywall"
  private val paywallWillOpenURL = "paywallWillOpenURL"
  private val paywallWillOpenDeepLink = "paywallWillOpenDeepLink"
  private val handleLog = "handleLog"
  private val willRedeemLink = "willRedeemLink"
  private val didRedeemLink = "didRedeemLink"
  val purchaseController = PurchaseControllerBridge.instance

  override fun definition() = ModuleDefinition {


    Name("SuperwallExpo")

    Events(
      onPaywallPresent,
      onPaywallDismiss,
      onPaywallError,
      onPaywallSkip,
      onCustomCallback,
      onPurchase,
      onPurchaseRestore,
      onBackPressed,

      // Legacy events
      subscriptionStatusDidChange,
      handleSuperwallEvent,
      handleCustomPaywallAction,
      willDismissPaywall,
      willPresentPaywall,
      didDismissPaywall,
      didPresentPaywall,
      paywallWillOpenURL,
      paywallWillOpenDeepLink,
      handleLog,
      willRedeemLink,
      didRedeemLink
    )

    AsyncFunction("identify") { userId: String, options: Map<String, Any>?, promise: Promise ->
      scope.launch {
        try {
          val identityOptions = options?.let { identityOptionsFromJson(it) }
          Superwall.instance.identify(userId = userId, options = identityOptions)
          promise.resolve(null)
        } catch (error: Exception) {
          promise.reject(CodedException(error))
        }
      }
    }

    AsyncFunction("reset") { promise: Promise ->
      scope.launch {
        try {
          Superwall.instance.reset()
          promise.resolve(null)
        } catch (error: Exception) {
          promise.reject(CodedException(error))
        }
      }
    }

    AsyncFunction("configure") {
      apiKey: String,
      options: Map<String, Any>?,
      usingPurchaseController: Boolean?,
      sdkVersion: String?,
      promise: Promise ->
      ioScope.launch {
        val promiseSettled = java.util.concurrent.atomic.AtomicBoolean(false)
        try{
        val superwallOptions: SuperwallOptions = options?.let {
          superwallOptionsFromJson(options)
        }?:SuperwallOptions()

        // Set up onBackPressed callback to emit event to JS and wait for response
        superwallOptions.paywalls.onBackPressed = { paywallInfo ->
          backPressedPromise = CompletableFuture()

          val data = paywallInfo?.let {
            mapOf("paywallInfo" to it.toJson())
          } ?: emptyMap()

          emitEvent("onBackPressed", data)

          // Block and wait for JS response via didHandleBackPressed
          runBlocking {
            backPressedPromise?.await() ?: false
          }
        }

        @Suppress("UNCHECKED_CAST")
        val localResourcesMap = options?.get("localResources") as? Map<String, String>

        Superwall.configure(
          apiKey = apiKey,
          applicationContext = appContext.reactContext?.applicationContext as Application,
          purchaseController = if (usingPurchaseController == true) purchaseController else null,
          activityProvider = ExpoActivityProvider(appContext),
          options = superwallOptions,
          completion = {
            Superwall.instance.setPlatformWrapper("Expo", version = sdkVersion ?: "0.0.0")
            Superwall.instance.delegate = SuperwallDelegateBridge()
            if (localResourcesMap != null) {
              Superwall.instance.localResources = localResourcesMap.mapValues { (_, value) ->
                PaywallResource.FromUri(Uri.parse(value))
              }
            }
            if (promiseSettled.compareAndSet(false, true)) {
              promise.resolve(true)
            }
           }
         )
        } catch (error: Throwable) {
          error.printStackTrace()
          if (promiseSettled.compareAndSet(false, true)) {
            scope.launch {
              promise.reject(CodedException(error))
            }
          }
        }
      }
    }

    AsyncFunction("getConfigurationStatus") { promise: Promise ->
      try {
        val configurationStatus = Superwall.instance.configurationState.asString()
          promise.resolve(configurationStatus)
        } catch (error: Throwable) {
          promise.reject(CodedException(error))
        }
      }

    AsyncFunction("registerPlacement") {
      placement: String,
      params: Map<String, Any>?,
      handlerId: String?,
      promise: Promise ->



      var handler: PaywallPresentationHandler? = handlerId?.let {
        PaywallPresentationHandler()
      }



        handler?.onPresent { paywallInfo ->
          val data =
          mapOf(
            "paywallInfoJson" to paywallInfo.toJson(),
          "handlerId" to handlerId)
          sendEvent(onPaywallPresent, data)
        }



        handler?.onDismiss { paywallInfo, result ->
          val data =
          mapOf(
            "paywallInfoJson" to paywallInfo.toJson(),
          "result" to result.toJson(),
          "handlerId" to handlerId)
          
          sendEvent(onPaywallDismiss, data)
        }

        handler?.onError { error ->
          val data =
          mapOf(
            "errorString" to error.localizedMessage,
          "handlerId" to handlerId,
          )
          sendEvent(onPaywallError, data)
        }

        handler?.onSkip { reason ->
          val data =
          mapOf(
            "skippedReason" to reason.toJson(),
             "handlerId" to handlerId)
          sendEvent(onPaywallSkip, data)
        }

        handler?.onCustomCallback { callback ->
          val callbackId = java.util.UUID.randomUUID().toString()
          val future = CompletableFuture<CustomCallbackResult>()
          customCallbackFutures[callbackId] = future

          val data = mutableMapOf<String, Any?>(
            "callbackId" to callbackId,
            "name" to callback.name,
            "handlerId" to handlerId,
          )
          callback.variables?.let { data["variables"] = it }

          sendEvent(onCustomCallback, data)

          future.await()
        }

      Superwall.instance.register(
        placement = placement,
        params = params,
        handler = handler
      ) {
        promise.resolve(null)
      }
    }
    

    AsyncFunction("getAssignments") { promise: Promise ->
        Superwall.instance.getAssignments()
        .fold({
          promise.resolve(it.map { it.toJson() })
        }, { error ->
          promise.reject(CodedException(error))
        })
      }

    AsyncFunction("getEntitlements") { promise: Promise ->
      try {
        val entitlements = Superwall.instance.entitlements
        val map = mutableMapOf<String, List<Map<String, Any>>>()

        map["all"] = entitlements.all.toJson()
        map["inactive"] = entitlements.inactive.toJson()
        map["active"] = entitlements.active.toJson()
        promise.resolve(map)
      } catch (error: Exception) {
        promise.reject(CodedException(error))
      }
      }

    AsyncFunction("getSubscriptionStatus") { promise: Promise ->
      try {
        val subscriptionStatus = Superwall.instance.subscriptionStatus.value.toJson()
          promise.resolve(subscriptionStatus)
      } catch (error: Exception) {
        promise.reject(CodedException(error))
      }
    }

    AsyncFunction("setSubscriptionStatus") { status: Map<String,Any>, promise: Promise ->
      scope.launch {
        try {
          val statusString = (status["status"] as? String)?.uppercase() ?: "UNKNOWN"
          val subscriptionStatus: SubscriptionStatus

          when (statusString) {
            "UNKNOWN" -> subscriptionStatus = SubscriptionStatus.Unknown
            "INACTIVE" -> subscriptionStatus = SubscriptionStatus.Inactive
            "ACTIVE" -> {
              val entitlementsArray = status["entitlements"] as? List<Map<String, Any>>
              val entitlementsSet: Set<Entitlement> = entitlementsArray?.map { item ->
                entitlementFromJson(item)
              }?.filterNotNull()?.toSet() ?: emptySet()
              subscriptionStatus = SubscriptionStatus.Active(entitlementsSet)
            }
            else -> subscriptionStatus = SubscriptionStatus.Unknown
          }

          Superwall.instance.setSubscriptionStatus(subscriptionStatus)
          promise.resolve(null)
        } catch (error: Exception) {
          promise.reject(CodedException(error))
        }
      }
    }

    Function("setInterfaceStyle") { style: String? ->
      var interfaceStyle: InterfaceStyle? = style?.let { interfaceStyleFromString(it) }
      Superwall.instance.setInterfaceStyle(interfaceStyle)
    }

    AsyncFunction("getUserAttributes") { promise: Promise ->
      val attributes = Superwall.instance.userAttributes
      promise.resolve(attributes)
    }

    AsyncFunction("getDeviceAttributes") { promise: Promise ->
      scope.launch {
        try {
            val attributes = Superwall.instance.deviceAttributes()
            promise.resolve(attributes)
        } catch (e: Exception) {
            promise.reject("GET_DEVICE_ATTRIBUTES_ERROR", e.message, e)
        }
      }
    }


    AsyncFunction("setUserAttributes") { userAttributes: Map<String, Any?>, promise: Promise ->
      scope.launch {
        try {
          Superwall.instance.setUserAttributes(userAttributes)
          promise.resolve(null)
        } catch (error: Exception) {
          promise.reject(CodedException(error))
        }
      }
    }

    AsyncFunction("handleDeepLink") { url: String, promise: Promise ->
      val url = Uri.parse(url)
      val result = Superwall.handleDeepLink(url).fold({
        promise.resolve(it)
      }, { error ->
        promise.resolve(false)
      })
    }

    Function("didPurchase") { result: Map<String, Any> ->
      val purchaseResult = purchaseResultFromJson(result)
      purchaseController.purchasePromise?.complete(purchaseResult)
    }

    Function("didRestore") { result: Map<String, Any> ->
      val restorationResult = restorationResultFromJson(result)
      purchaseController.restorePromise?.complete(restorationResult)
    }

    Function("didHandleBackPressed") { shouldConsume: Boolean ->
      backPressedPromise?.complete(shouldConsume)
    }

    AsyncFunction("didHandleCustomCallback") { callbackId: String, status: String, data: Map<String, Any>?, promise: Promise ->
      val future = customCallbackFutures.remove(callbackId)
      if (future == null) {
        promise.resolve(null)
        return@AsyncFunction
      }
      val result = if (status == "success") {
        CustomCallbackResult.success(data)
      } else {
        CustomCallbackResult.failure(data)
      }
      future.complete(result)
      promise.resolve(null)
    }

    AsyncFunction("restorePurchases") { promise: Promise ->
      ioScope.launch {
        Superwall.instance.restorePurchases().fold({ result ->
          scope.launch {
            promise.resolve(restorationResultToJson(result))
          }
        }, { error ->
          scope.launch {
            promise.resolve(restorationResultToJson(RestorationResult.Failed(error)))
          }
        })
      }
    }

    AsyncFunction("dismiss") { promise: Promise ->
      ioScope.launch {
        Superwall.instance.dismiss()
        scope.launch {
          promise.resolve(null)
        }
      }
    }

    AsyncFunction("confirmAllAssignments") { promise: Promise ->
      ioScope.launch {
        Superwall.instance.confirmAllAssignments().fold({
          scope.launch {
            promise.resolve(it.map { it.toJson() })
          }
        }, { error ->
          scope.launch {
            promise.reject(CodedException(error))
          }
        })
      }
    }

    AsyncFunction("getPresentationResult") { placement: String, params: Map<String, Any>?, promise: Promise ->
      ioScope.launch {
        Superwall.instance.getPresentationResult(placement = placement, params = params)
        .fold({ result ->
          scope.launch {
            promise.resolve(result.toJson())
          }
        }, { error ->
          scope.launch {
            promise.reject(CodedException(error))
          }
        })
      }
    }

    Function("preloadPaywalls") { placementNames: List<String> ->
      Superwall.instance.preloadPaywalls(placementNames = placementNames.toSet())
    }

    Function("preloadAllPaywalls") {
      Superwall.instance.preloadAllPaywalls()
    }

    Function("setLogLevel") { level: String ->
      val logLevel = LogLevel.values().find { it.toString().equals(level, ignoreCase = true) }
      if (logLevel != null) {
        Superwall.instance.logLevel = logLevel
      }
    }

    Function("setEventTrackingBehavior") { behavior: String ->
      when (behavior) {
        "all" -> Superwall.instance.eventTrackingBehavior = EventTrackingBehavior.ALL
        "superwallOnly" -> Superwall.instance.eventTrackingBehavior = EventTrackingBehavior.SUPERWALL_ONLY
        "none" -> Superwall.instance.eventTrackingBehavior = EventTrackingBehavior.NONE
      }
    }

    AsyncFunction("setIntegrationAttributes") { attributes: Map<String, String>, promise: Promise ->
      scope.launch {
        try {
          val converted = attributes.mapNotNull { (key, value) ->
            attributionProviderFromString(key)?.let { it to value }
          }.toMap()

          Superwall.instance.setIntegrationAttributes(converted)
          promise.resolve(null)
        } catch (error: Exception) {
          promise.reject(CodedException(error))
        }
      }
    }

    AsyncFunction("getIntegrationAttributes") { promise: Promise ->
      val attributes = Superwall.instance.integrationAttributes
      promise.resolve(attributes)
    }

    AsyncFunction("consume") { purchaseToken: String, promise: Promise ->
      ioScope.launch {
        Superwall.instance.consume(purchaseToken).fold({ token ->
          scope.launch { promise.resolve(token) }
        }, { error ->
          scope.launch { promise.reject(CodedException(error)) }
        })
      }
    }
  }
}

File: ios/SuperwallExpoModule.swift
import ExpoModulesCore
import SuperwallKit

/// Thread-safe boolean flag that can only be set once.
private final class AtomicFlag {
  private var _value = false
  private let lock = NSLock()

  /// Sets the flag and returns the previous value.
  /// Returns `false` on the first call, `true` on subsequent calls.
  func testAndSet() -> Bool {
    lock.lock()
    defer { lock.unlock() }
    let prev = _value
    _value = true
    return prev
  }
}

public class SuperwallExpoModule: Module {
  public static var shared: SuperwallExpoModule?
  private static let instancesLock = NSLock()
  private static let instances = NSHashTable<SuperwallExpoModule>.weakObjects()

  private let purchaseController = PurchaseControllerBridge.shared
  private var delegate: SuperwallDelegateBridge?

  // Paywall Events
  private let onPaywallPresent = "onPaywallPresent"
  private let onPaywallDismiss = "onPaywallDismiss"
  private let onPaywallError = "onPaywallError"
  private let onPaywallSkip = "onPaywallSkip"
  private let onCustomCallback = "onCustomCallback"

  // Custom callback continuations keyed by callbackId
  private let callbackLock = NSLock()
  private var customCallbackContinuations: [String: CheckedContinuation<CustomCallbackResult, Never>] = [:]

  // Purchase Events
  private let onPurchase = "onPurchase"
  private let onPurchaseRestore = "onPurchaseRestore"

  // Legacy Events
  private let subscriptionStatusDidChange = "subscriptionStatusDidChange"
  private let handleSuperwallEvent = "handleSuperwallEvent"
  private let handleCustomPaywallAction = "handleCustomPaywallAction"
  private let willDismissPaywall = "willDismissPaywall"
  private let willPresentPaywall = "willPresentPaywall"
  private let didDismissPaywall = "didDismissPaywall"
  private let didPresentPaywall = "didPresentPaywall"
  private let paywallWillOpenURL = "paywallWillOpenURL"
  private let paywallWillOpenDeepLink = "paywallWillOpenDeepLink"
  private let handleLog = "handleLog"
  private let willRedeemLink = "willRedeemLink"
  private let didRedeemLink = "didRedeemLink"

  public required init(appContext: AppContext) {
    super.init(appContext: appContext)
    SuperwallExpoModule.shared = self
    SuperwallExpoModule.instancesLock.lock()
    SuperwallExpoModule.instances.add(self)
    SuperwallExpoModule.instancesLock.unlock()
  }

  // Events must reach every live module instance. More than one app context can
  // exist at once (e.g. expo-dev-client's launcher plus the app), and `shared`
  // may point at an instance whose JS runtime never subscribed - sending only
  // through it silently drops events like `onPurchase`, leaving the native
  // purchase continuation suspended forever.
  public static func emitEvent(_ name: String, _ data: [String: Any]?) {
    instancesLock.lock()
    let liveInstances = instances.allObjects
    instancesLock.unlock()
    for instance in liveInstances {
      instance.sendEvent(name, data ?? [:])
    }
  }

  public func definition() -> ModuleDefinition {
    Name("SuperwallExpo")

    Events(
      onPaywallPresent,
      onPaywallDismiss,
      onPaywallError,
      onPaywallSkip,
      onCustomCallback,
      onPurchase,
      onPurchaseRestore,

      // Legacy events
      subscriptionStatusDidChange,
      handleSuperwallEvent,
      handleCustomPaywallAction,
      willDismissPaywall,
      willPresentPaywall,
      didDismissPaywall,
      didPresentPaywall,
      paywallWillOpenURL,
      paywallWillOpenDeepLink,
      handleLog,
      willRedeemLink,
      didRedeemLink
    )

    View(SuperwallExpoPaywallView.self) {
      ViewName("PaywallView")
      Events(
        "onPaywallLoadStart",
        "onPaywallPresent",
        "onPaywallDismiss",
        "onPaywallSkip",
        "onPaywallError"
      )
      Prop("placement") { (view: SuperwallExpoPaywallView, placement: String) in
        view.setPlacement(placement)
      }
      Prop("params") { (view: SuperwallExpoPaywallView, params: [String: Any]?) in
        view.setParams(params)
      }
      OnViewDidUpdateProps { (view: SuperwallExpoPaywallView) in
        view.loadPaywallIfNeeded()
      }
    }

    AsyncFunction("identify") { (userId: String, options: [String: Any]?, promise: Promise) in
      DispatchQueue.main.async {
        let identityOptions = IdentityOptions.fromJson(options)
        Superwall.shared.identify(userId: userId, options: identityOptions)
        promise.resolve(nil)
      }
    }

    AsyncFunction("reset") { (promise: Promise) in
      DispatchQueue.main.async {
        Superwall.shared.reset()
        promise.resolve(nil)
      }
    }

    AsyncFunction("configure") {
      (
        apiKey: String,
        options: [String: Any]?,
        usingPurchaseController: Bool,
        sdkVersion: String?,
        promise: Promise
      ) in

      var superwallOptions: SuperwallOptions?

      if let options = options {
        superwallOptions = SuperwallOptions.fromJson(options)
      }

      let promiseSettled = AtomicFlag()

      Superwall.configure(
        apiKey: apiKey,
        purchaseController: usingPurchaseController ? purchaseController : nil,
        options: superwallOptions,
        completion: {
          self.delegate = SuperwallDelegateBridge()
          Superwall.shared.delegate = self.delegate

          Superwall.shared.setPlatformWrapper("Expo", version: sdkVersion ?? "0.0.0")

          if promiseSettled.testAndSet() == false {
            promise.resolve(nil)
          }
        }
      )
    }

    AsyncFunction("getConfigurationStatus") { (promise: Promise) in
      let configurationStatus = Superwall.shared.configurationStatus.toString()
      promise.resolve(configurationStatus)
    }

    AsyncFunction("registerPlacement") {
      (
        placement: String,
        params: [String: Any]?,
        handlerId: String?,
        promise: Promise
      ) in
      var handler: PaywallPresentationHandler?

      if let handlerId = handlerId {
        handler = PaywallPresentationHandler()

        handler?.onPresent { [weak self] paywallInfo in
          guard let self = self else { return }
          let data =
            [
              "paywallInfoJson": paywallInfo.toJson(),
              "handlerId": handlerId,
            ] as [String: Any]
          self.sendEvent(self.onPaywallPresent, data)
        }

        handler?.onDismiss { [weak self] paywallInfo, result in
          guard let self = self else { return }
          let data =
            [
              "paywallInfoJson": paywallInfo.toJson(),
              "result": result.toJson(),
              "handlerId": handlerId,
            ] as [String: Any]
          self.sendEvent(self.onPaywallDismiss, data)
        }

        handler?.onError { [weak self] error in
          guard let self = self else { return }
          let data =
            [
              "errorString": error.localizedDescription,
              "handlerId": handlerId,
            ] as [String: Any]
          self.sendEvent(self.onPaywallError, data)
        }

        handler?.onSkip { [weak self] reason in
          guard let self = self else { return }
          let data =
            [
              "skippedReason": reason.toJson(),
              "handlerId": handlerId,
            ] as [String: Any]
          self.sendEvent(self.onPaywallSkip, data)
        }

        handler?.onCustomCallback { [weak self] callback async -> CustomCallbackResult in
          guard let self = self else { return .failure() }
          let callbackId = UUID().uuidString
          let data: [String: Any] = [
            "callbackId": callbackId,
            "name": callback.name,
            "variables": callback.variables as Any,
            "handlerId": handlerId,
          ]
          return await withCheckedContinuation { continuation in
            self.callbackLock.lock()
            self.customCallbackContinuations[callbackId] = continuation
            self.callbackLock.unlock()
            self.sendEvent(self.onCustomCallback, data)
          }
        }
      }

      Superwall.shared.register(
        placement: placement,
        params: params,
        handler: handler
      ) {
        promise.resolve(nil)
      }
    }

    AsyncFunction("getAssignments") { (promise: Promise) in
      do {
        let assignments = try Superwall.shared.getAssignments()
        promise.resolve(assignments.map { $0.toJson() })
      } catch {
        promise.reject(error)
      }
    }

    AsyncFunction("getEntitlements") { (promise: Promise) in
      do {
        let entitlements = try Superwall.shared.entitlements.toJson()
        promise.resolve(entitlements)
      } catch {
        promise.reject(error)
      }
    }

    AsyncFunction("getSubscriptionStatus") { (promise: Promise) in
      print("Getting subscription status")
      let subscriptionStatus = Superwall.shared.subscriptionStatus
      promise.resolve(subscriptionStatus.toJson())
    }

    AsyncFunction("setSubscriptionStatus") { (status: [String: Any], promise: Promise) in
      DispatchQueue.main.async {
        let statusString = (status["status"] as? String)?.uppercased() ?? "UNKNOWN"
        let subscriptionStatus: SubscriptionStatus

        print("Setting subscription status to: \(statusString)")

        switch statusString {
        case "UNKNOWN":
          subscriptionStatus = .unknown
        case "INACTIVE":
          subscriptionStatus = .inactive
        case "ACTIVE":
          if let entitlementsArray = status["entitlements"] as? [[String: Any]] {
            let entitlementsSet: Set<Entitlement> = Set(
              entitlementsArray.compactMap { item in
                if let id = item["id"] as? String {
                  return Entitlement(id: id)
                }
                return nil
              }
            )
            subscriptionStatus = .active(entitlementsSet)
          } else {
            subscriptionStatus = .inactive
          }
        default:
          subscriptionStatus = .unknown
        }

        Superwall.shared.subscriptionStatus = subscriptionStatus
        promise.resolve(nil)
      }
    }

    Function("setInterfaceStyle") { (style: String?) in
      var interfaceStyle: InterfaceStyle?
      if let style = style {
        interfaceStyle = InterfaceStyle.fromString(style: style)
      }
      Superwall.shared.setInterfaceStyle(to: interfaceStyle)
    }

    AsyncFunction("getUserAttributes") { (promise: Promise) in
      let attributes = Superwall.shared.userAttributes
      promise.resolve(attributes)
    }

    AsyncFunction("getDeviceAttributes") {
      return await Superwall.shared.getDeviceAttributes()
    }

    AsyncFunction("setUserAttributes") { (userAttributes: [String: Any?], promise: Promise) in
      DispatchQueue.main.async {
        Superwall.shared.setUserAttributes(userAttributes)
        promise.resolve(nil)
      }
    }

    AsyncFunction("handleDeepLink") { (url: String, promise: Promise) in
      guard let url = URL(string: url) else {
        promise.resolve(false)
        return
      }
      let result = Superwall.shared.handleDeepLink(url)
      promise.resolve(result)
    }

    Function("didPurchase") { (result: [String: Any]) in
      guard let purchaseResult = PurchaseResult.fromJson(result) else {
        return
      }
      purchaseController.purchaseCompletion?(purchaseResult)
    }

    Function("didRestore") { (result: [String: Any]) in
      guard let restorationResult = RestorationResult.fromJson(result) else {
        return
      }
      purchaseController.restoreCompletion?(restorationResult)
    }

    AsyncFunction("didHandleCustomCallback") { (callbackId: String, status: String, data: [String: Any]?, promise: Promise) in
      self.callbackLock.lock()
      let continuation = self.customCallbackContinuations.removeValue(forKey: callbackId)
      self.callbackLock.unlock()
      guard let continuation = continuation else {
        promise.resolve(nil)
        return
      }
      let result: CustomCallbackResult
      if status == "success" {
        result = .success(data: data)
      } else {
        result = .failure(data: data)
      }
      continuation.resume(returning: result)
      promise.resolve(nil)
    }

    AsyncFunction("restorePurchases") { (promise: Promise) in
      Task {
        let result = await Superwall.shared.restorePurchases()
        promise.resolve(result.toJson())
      }
    }

    AsyncFunction("dismiss") { (promise: Promise) in
      Superwall.shared.dismiss {
        promise.resolve(nil)
      }
    }

    AsyncFunction("confirmAllAssignments") { (promise: Promise) in
      Superwall.shared.confirmAllAssignments { assignments in
        promise.resolve(assignments.map { $0.toJson() })
      }
    }

    AsyncFunction("getPresentationResult") {
      (placement: String, params: [String: Any]?, promise: Promise) in
      Superwall.shared.getPresentationResult(forPlacement: placement, params: params) { result in
        promise.resolve(result.toJson())
      }
    }

    Function("preloadPaywalls") { (placementNames: [String]) in
      Superwall.shared.preloadPaywalls(forPlacements: Set(placementNames))
    }

    Function("preloadAllPaywalls") {
      Superwall.shared.preloadAllPaywalls()
    }

    Function("setLogLevel") { (level: String) in
      let logLevel = LogLevel.fromJson(level)
      if let logLevel = logLevel {
        Superwall.shared.logLevel = logLevel
      }
    }

    Function("setEventTrackingBehavior") { (behavior: String) in
      switch behavior {
      case "all": Superwall.shared.eventTrackingBehavior = .all
      case "superwallOnly": Superwall.shared.eventTrackingBehavior = .superwallOnly
      case "none": Superwall.shared.eventTrackingBehavior = .none
      default: break
      }
    }

    AsyncFunction("setIntegrationAttributes") { (attributes: [String: String], promise: Promise) in
      var converted: [IntegrationAttribute: String] = [:]

      for (key, value) in attributes {
        if let attribute = IntegrationAttribute.fromString(key) {
          converted[attribute] = value
        }
      }

      Superwall.shared.setIntegrationAttributes(converted)
      promise.resolve(nil)
    }

    AsyncFunction("getIntegrationAttributes") { (promise: Promise) in
      let attributes = Superwall.shared.integrationAttributes
      promise.resolve(attributes)
    }

    AsyncFunction("consume") { (_: String, promise: Promise) in
      promise.resolve(nil)
    }
  }
}
